File: dbusmenu4j/src/main/java/dev/jfronny/dbusmenu4j/Menu.java
package dev.jfronny.dbusmenu4j;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;

public interface Menu {
    int getId();
    boolean isSeparator();
    @NotNull String getLabel();
    boolean isEnabled();
    boolean isVisible();
    byte @Nullable [] getIconData();
    String @Nullable [] getShortcut();
    @Nullable String getToggleType();
    int getToggleState();
    @Nullable List<Menu> getChildren();

    void onEvent();
    void update();
    void maybeUpdate();

    abstract class Abstract implements Menu {
        private long lastUpdated = 0;

        @Override
        public void maybeUpdate() {
            if (System.currentTimeMillis() - lastUpdated > 1000) update();
        }

        @Override
        public void update() {
            lastUpdated = System.currentTimeMillis();
        }
    }
}

File: src/main/kotlin/io/gitlab/jfronny/globalmenu/proxy/Menu.kt
package io.gitlab.jfronny.globalmenu.proxy

interface Menu {
    val id: Int
    val isSeparator: Boolean
    val label: String
    val isEnabled: Boolean
    val isVisible: Boolean
    val iconData: ByteArray?
    val shortcut: Array<String>?
    val toggleType: String?
    val toggleState: Int
    val children: List<Menu>?
    fun onEvent()
    fun update()
    fun maybeUpdate()

    abstract class Abstract : Menu {
        private var lastUpdated = 0L
        override fun maybeUpdate() {
            if (System.currentTimeMillis() - lastUpdated > 1000) {
                update()
            }
        }

        override fun update() {
            lastUpdated = System.currentTimeMillis()
        }
    }
}